crates/services/sync/src/state.rs

```diff
@@ -0,0 +1,212 @@
+                    // If the failure range contains the start of the processing range,
+                    // then there is no reason to continue trying to process this range.
+                    // The processing range is reverted back to just before it's start.
+                    // The revert is either to the last committed height, or to uninitialized.
+                    .contains(processing.start())
+                    .then(|| {
+                        processing
+                            .start()
+                            .checked_sub(1)
+                            .map_or(Status::Uninitialized, Status::Committed)
+                    })
+                    .or_else(|| {
+                        // If the failure range contains the end of the processing range,
+                        // or the processing range contains the start of the failure range,
+                        // then the processing range is shortened to just before the failure range.
+                        (range.contains(processing.end())
+                            || processing.contains(range.start()))
+                        .then(|| {
+                            range
+                                .start()
+                                .checked_sub(1)
+                                .map_or(Status::Uninitialized, |prev| {
+                                    Status::Processing(*processing.start()..=prev)
+                                })
+                        })
+                    })
+                    .or_else(|| {
+                        // If the processing range contains the end of the failure range,
+                        // then the entire processing range is failed and reverted back to
+                        // the last committed height, or to uninitialized.
+                        processing.contains(range.end()).then(|| {
+                            processing
+                                .start()
+                                .checked_sub(1)
+                                .map_or(Status::Uninitialized, Status::Committed)
+                        })
+                    }),
+            });
+        self.apply_status(status);
+    }
+
+    /// Get the current observed height.
+    pub fn proposed_height(&self) -> Option<&u32> {
+        match &self.status {
+            Status::Processing(range) => Some(range.end()),
+            _ => None,
+        }
+    }
+
+    fn apply_status(&mut self, status: Option<Status>) {
+        if let Some(s) = status {
+            self.status = s;
+        }
+    }
+}
+
+/// If a commit is made to a height that is
+/// below the existing committed height this is
+/// new evidence and we check if there is a gap between
+/// the existing committed height and the new commit.
+///
+/// This case should not occur but because we must handle
+/// it then the most resilient way is to assume that we
+/// should re-process the gap.
+fn commit_creates_processing(
```

Comment:
So this is a bit of a hard one. Really by design we can't ever have this happen because the commits are run in order and there's no state that can lead to re-running a commit. However I can't handle this case as an error at the stream level because there's nothing you can do so for the sake of completeness I needed to do something here. If we get a commit that is lower then the current commit then something is wrong and my way of looking at this is we can't trust the higher commit value so it should reprocess the headers in between. But you could make an argument either way.